trio/_threads.py

```diff
@@ -240,37 +315,64 @@ def deliver_worker_fn_result(result: outcome.Outcome[RetT]) -> None:
         limiter.release_on_behalf_of(placeholder)
         raise
 
-    def abort(_: RaiseCancelT) -> trio.lowlevel.Abort:
+    def abort(raise_cancel: RaiseCancelT) -> trio.lowlevel.Abort:
+        # fill so from_thread_check_cancelled can raise
+        cancel_register[0] = raise_cancel
         if cancellable:
+            # empty so report_back_in_trio_thread_fn cannot reschedule
             task_register[0] = None
             return trio.lowlevel.Abort.SUCCEEDED
         else:
             return trio.lowlevel.Abort.FAILED
 
-    # wait_task_rescheduled return value cannot be typed
-    return await trio.lowlevel.wait_task_rescheduled(abort)  # type: ignore[no-any-return]
+    while True:
+        # wait_task_rescheduled return value cannot be typed
+        msg_from_thread: ThreadDone[RetT] | Run[object] | RunSync[
+            object
+        ] = await trio.lowlevel.wait_task_rescheduled(abort)
+        if type(msg_from_thread) is ThreadDone:
+            return msg_from_thread.result.unwrap()  # type: ignore[no-any-return]
+        elif type(msg_from_thread) is Run:
+            await msg_from_thread.run()
+        elif type(msg_from_thread) is RunSync:
+            msg_from_thread.run_sync()
+        else:  # pragma: no cover, internal debugging guard
+            raise TypeError(
+                "trio.to_thread.run_sync received unrecognized thread message {!r}."
+                "".format(msg_from_thread)
+            )
+        del msg_from_thread
 
 
-# We use two typevars here, because cb can transform from one to the other any way it likes.
-def _run_fn_as_system_task(
-    cb: Callable[
-        [
-            stdlib_queue.SimpleQueue[outcome.Outcome[Ret2T]],
-            Callable[..., RetT],
-            tuple[object, ...],
-        ],
-        object,
-    ],
-    fn: Callable[..., RetT],
-    *args: object,
-    context: contextvars.Context,
-    trio_token: TrioToken | None = None,
-    # Outcome isn't typed, so Ret2T is used only in the return type.
-) -> Ret2T:  # type: ignore[type-var]
-    """Helper function for from_thread.run and from_thread.run_sync.
+def from_thread_check_cancelled() -> None:
+    """Raise trio.Cancelled if the associated Trio task entered a cancelled status.
+
+     Only applicable to threads spawned by `trio.to_thread.run_sync`. Poll to allow
+     ``cancellable=False`` threads to raise :exc:`trio.Cancelled` at a suitable
```

Comment:
man, we really need a better name for this `cancellable` thing now. Not something that needs to be changed in this PR, but maybe we should mention the history in the docs for `to_thread.run_sync`.